From 87281952c852f3b44431c74d3aad5222e67d0be6 Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Tue, 21 Nov 2023 21:35:42 +0530
Subject: [PATCH 1/2] wifi: mac80211: handle invalid link id in
 ieee80211_rx_data_set_sta

mac80211:
The current logic in ieee80211_rx_data_set_sta to handle link_id < 0
doesn't include the non-eht station case.

link_id is expected to be less than 0 for non-eht station.

Add support to handle the above case.

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 net/mac80211/rx.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -4424,8 +4424,16 @@ static bool ieee80211_rx_data_set_sta(st
 	}
 
 	if (link_id < 0) {
-		rx->link = &rx->sdata->deflink;
-		WARN_ON(!!rx->sdata->vif.valid_links);
+		if (rx->sdata->vif.valid_links) {
+			if (sta && !sta->sta.valid_links) {
+				rx->link = rcu_dereference(rx->sdata->link[sta->deflink.link_id]);
+			} else {
+				rx->link = &rx->sdata->deflink;
+				WARN_ON_ONCE(1);
+			}
+		} else {
+			rx->link = &rx->sdata->deflink;
+		}
 	}
 	else if (!ieee80211_rx_data_set_link(rx, link_id))
 		return false;
